File: jcv-core/src/main/java/com/ekino/oss/jcv/core/validator/ValidatorTemplateManager.java
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core.validator;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

import lombok.RequiredArgsConstructor;

import static java.util.Optional.*;

/**
 * Manager of templated validator format.
 * Example : {@code {#some_id:first parameter;the next \; parameter;the last one...#}}
 *
 * @author Leo Millon
 */
@RequiredArgsConstructor
public class ValidatorTemplateManager {

    private static final Pattern VALIDATOR_TEMPLATE_PATTERN = Pattern.compile("^\\{#(.+)#}$");
    private static final Pattern PARAMETERS_PATTERN = Pattern.compile("^(?<id>[\\w-_.]+)(:(?<parameters>(([^;]+))?(;([^;])+)*))?$");

    private final String value;
    private String extractedTemplate;

    /**
     * A method to test if the template format is valid.
     *
     * @return true if the template format is valid
     */
    public boolean isValidTemplate() {
        return extractTemplateContent() != null;
    }

    private Optional<Matcher> templateMatcher() {
        return ofNullable(value)
            .map(VALIDATOR_TEMPLATE_PATTERN::matcher);
    }

    /**
     * Extract the validator definition.
     *
     * For {@code {#some_id:first parameter#}}, the content will be {@code some_id:first parameter#}
     *
     * @return the template content, or null (if format is invalid)
     */
    public String extractTemplateContent() {

        if (extractedTemplate == null) {
            extractedTemplate = templateMatcher()
                .filter(Matcher::matches)
                .map(it -> it.group(1))
                .orElse(null);
        }

        return extractedTemplate;
    }

    /**
     * Extract the validator id.
     *
     * For {@code {#some_id:first parameter#}}, the validator id will be {@code some_id}.
     *
     * @return the validator id, or null (if format is invalid)
     */
    public String extractId() {
        return parameterMatcher()
            .filter(Matcher::matches)
            .map(it -> it.group("id"))
            .orElse(null);
    }

    /**
     * Extract the validator parameters.
     *
     * For {@code {#some_id:first parameter;the next \; parameter;the last one...#}}, the parameters will be :
     * <ul>
     * <li>first parameter</li>
     * <li>the next ; parameter</li>
     * <li>the last one...</li>
     * </ul>
     *
     * @return the ordered list of parameters, or empty (if format is invalid)
     */
    public List<String> extractParameters() {
        return parameterMatcher()
            .filter(Matcher::matches)
            .map(it -> it.group("parameters"))
            .map(it -> it.split("(?<!\\\\);"))
            .map(Arrays::asList)
            .map(espacedParam -> espacedParam.stream()
                .map(it -> it.replaceAll("\\\\;", ";"))
                .collect(Collectors.toList()))
            .orElseGet(Collections::emptyList);
    }

    /**
     * Extract the validator parameters.
     *
     * For {@code {#some_id:first parameter;the next \; parameter;the last one...#}}, the parameter {@code 0} will be {@code first parameter}.
     *
     * @param index the (0-based) index
     *
     * @return the parameter at index (0-based index), or empty (if format is invalid or out of bound)
     */
    public Optional<String> extractParameter(int index) {
        List<String> parameters = extractParameters();

        return of(index)
            .filter(it -> it >= 0)
            .filter(it -> it < parameters.size())
            .map(parameters::get);
    }

    private Optional<Matcher> parameterMatcher() {
        return ofNullable(extractTemplateContent())
            .map(PARAMETERS_PATTERN::matcher);
    }
}

File: jcv-core/src/main/kotlin/com/ekino/oss/jcv/core/validator/ValidatorTemplateManager.kt
/*
 * Copyright (c) 2019 ekino (https://www.ekino.com/)
 */
package com.ekino.oss.jcv.core.validator

/**
 * Manager of templated validator format.
 * Example : `{#some_id:first parameter;the next \; parameter;the last one...#}`
 *
 * @author Leo Millon
 */
class ValidatorTemplateManager(private val value: String?) {

    companion object {
        private val VALIDATOR_TEMPLATE_REGEX = "^\\{#(.+)#}$".toRegex()
        private val PARAMETERS_REGEX = "^(?<id>[\\w-_.]+)(:(?<parameters>([^;]+)?(;([^;])+)*))?$".toRegex()
        private val PARAMETER_SEPARATOR_REGEX = "(?<!\\\\);".toRegex()
    }

    private val extractedTemplate: String? by lazy {
        value?.let { VALIDATOR_TEMPLATE_REGEX.find(it) }?.groups?.get(1)?.value
    }

    /**
     * A method to test if the template format is valid.
     *
     * @return true if the template format is valid
     */
    val isValidTemplate: Boolean
        get() = extractTemplateContent() != null

    /**
     * Extract the validator definition.
     *
     * For `{#some_id:first parameter#}`, the content will be `some_id:first parameter#`
     *
     * @return the template content, or null (if format is invalid)
     */
    fun extractTemplateContent(): String? = extractedTemplate

    /**
     * Extract the validator id.
     *
     * For `{#some_id:first parameter#}`, the validator id will be `some_id`.
     *
     * @return the validator id, or null (if format is invalid)
     */
    fun extractId(): String? = extractedTemplate
        .extractRegexGroup(PARAMETERS_REGEX, "id")

    /**
     * Extract the validator parameters.
     *
     * For `{#some_id:first parameter;the next \; parameter;the last one...#}`, the parameters will be :
     *
     *  * first parameter
     *  * the next ; parameter
     *  * the last one...
     *
     *
     * @return the ordered list of parameters, or empty (if format is invalid)
     */
    fun extractParameters(): List<String> = extractedTemplate
        .extractRegexGroup(PARAMETERS_REGEX, "parameters")
        ?.split(PARAMETER_SEPARATOR_REGEX)
        ?.dropLastWhile { it.isEmpty() }
        ?.map { it.replace("\\\\;".toRegex(), ";") }
        .orEmpty()

    /**
     * Extract the validator parameters.
     *
     * For `{#some_id:first parameter;the next \; parameter;the last one...#}`, the parameter `0` will be `first parameter`.
     *
     * @param index the (0-based) index
     *
     * @return the parameter at index (0-based index), or empty (if format is invalid or out of bound)
     */
    fun extractParameter(index: Int): String? {
        val parameters = extractParameters()

        return if (index in parameters.indices) parameters[index] else null
    }

    /**
     * Fallback method to handle regex group extraction as the stdlib-jdk8 does not support named group on MatchResult after JDK8.
     *
     * @see <a href="https://youtrack.jetbrains.com/issue/KT-20865">Retrieving groups by name is not supported on Java 9 even with `kotlin-stdlib-jre8` in the classpath</a>
     */
    private fun String?.extractRegexGroup(regex: Regex, groupName: String): String? {
        return this
            ?.let { regex.toPattern().matcher(it) }
            ?.takeIf { it.matches() }
            ?.group(groupName)
    }
}
